### PR TITLE
feat(server-kafka): add internal kafka metrics

### PR DESCRIPTION
**Pull request by sliesche**

### CODE REVIEW

File: sda-commons-server-prometheus/README.md [hunk 1]
```diff
@@ -34,6 +34,97 @@ Default metrics that are provided at `/metrics/prometheus`:
 *) A filter that extracts the consumer from the HTTP headers should add `Consumer-Name` to the request properties. That
    filter is not part of the `PrometheusBundle`.
 
+
+Kafka internal producer metrics have been exposed
+
+| Metric name                                                | Labels                 | Description                                                                                                                            | Source |
+|------------------------------------------------------------|------------------------|----------------------------------------------------------------------------------------------------------------------------------------|--------|
+| `kafka_producer_batch_split_rate`                          | `client.id`            | The average number of batch splits per second                                                                                          |        |
```
Comment (elbiocaetano) [inline sda-commons-server-prometheus/README.md:42]: Is it possible to write a test that validates all these metrics are being sent, or it is not possible / too hard in a test context?

Comment (JoergSiebahn) [reply sda-commons-server-prometheus/README.md:42]: That would be nice to ensure backward compatibility. Or at least notice when something changes.

Comment (sliesche) [reply sda-commons-server-prometheus/README.md:42]: Test added to KafkaPrometheusMonitoringIT

File: sda-commons-server-kafka/src/test/java/org/sdase/commons/server/kafka/KafkaPrometheusMonitoringIT.java [hunk 2]
```diff
@@ -140,4 +137,197 @@ void shouldWriteHelpAndTypeToMetrics() {
...
+            .collect(Collectors.toList());
+
+    assertThat(metricNamesMicrometer).containsAll(getMetricsThatShouldExist());
+  }
+
+  private List<String> getMetricsThatShouldExist() {
+
+    ArrayList<String> result = new ArrayList<>();
+
+    result.add("kafka_producer_batch_split_rate");
```
Comment (elbiocaetano) [inline sda-commons-server-kafka/src/test/java/org/sdase/commons/server/kafka/KafkaPrometheusMonitoringIT.java:221]: As a suggestion, you can use List.of("string1", "string2", ...)

**Review by elbiocaetano — APPROVED** [review]

File: sda-commons-server-kafka/src/test/java/org/sdase/commons/server/kafka/KafkaPrometheusMonitoringIT.java (hunk 2 above)
Comment (sliesche) [reply sda-commons-server-kafka/src/test/java/org/sdase/commons/server/kafka/KafkaPrometheusMonitoringIT.java]: updated

File: sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java [hunk 3]
```diff
@@ -60,10 +62,12 @@
 
   private final List<MessageListener<?, ?>> messageListeners = new ArrayList<>();
   private final List<ThreadedMessageListener<?, ?>> threadedMessageListeners = new ArrayList<>();
-  private final List<KafkaMessageProducer<?, ?>> messageProducers = new ArrayList<>();
+  private final Map<String, KafkaMessageProducer<?, ?>> messageProducers = new HashMap<>();
```
Comment (JoergSiebahn) [inline sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java:65]: Can we stay with `List` here? It seems like the grouping by `clientId` is not needed.

Comment (sliesche) [reply sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java:65]: I changed that to be able to remove metrics when producers are stoped.
see line 458

Comment (JoergSiebahn) [reply sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java:65]: Ah … I looked for the usage but overlooked the `k`.

File: sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java [hunk 4]
```diff
@@ -443,9 +452,10 @@ private void setupManagedThreadManager(Environment environment) {
 
   private void stopProducers() {
     messageProducers.forEach(
-        p -> {
+        (k, v) -> {
           try {
-            p.close();
+            v.close();
+            micrometerProducerListener.producerRemoved(k);
```
Comment (JoergSiebahn) [inline sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java:458]: Not sure, how important this is … but if it is really important, this call might be moved in a finally block.

File: sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/prometheus/MicrometerProducerListener.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import io.micrometer.core.instrument.binder.kafka.KafkaClientMetrics;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.kafka.clients.producer.Producer;
+
+/** Listener adds or removes Kafka internal metrics for each producer to a micrometer registry. */
+public class MicrometerProducerListener {
```
Comment (JoergSiebahn) [inline sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/prometheus/MicrometerProducerListener.java:14]: Do we want to move this class one level up and make it package private provide to not expose it as public API.

File: sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java (hunk 4 above)
Comment (sliesche) [reply sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java]: moved to finally

**Review by JoergSiebahn — APPROVED** [review]

**Pull request merged**